crates/turborepo-filewatch/src/package_watcher.rs

```diff
@@ -0,0 +1,219 @@
+        discovery: T,
+    ) -> Result<Self, package_manager::Error> {
+        let (exit_tx, exit_rx) = oneshot::channel();
+        let package_data: Arc<Mutex<_>> = Default::default();
+        let subscriber = Subscriber::new(exit_rx, root, package_data.clone(), recv, discovery)?;
+        let handle = tokio::spawn(subscriber.watch());
+        Ok(Self {
+            _exit_tx: exit_tx,
+            _handle: handle,
+            package_data,
+        })
+    }
+
+    pub async fn get_package_data(&self) -> Vec<RootWorkspaceData> {
+        self.package_data
+            .lock()
+            .expect("not poisoned")
+            .clone()
+            .into_values()
+            .collect()
+    }
+}
+
+/// The underlying task that listens to file system events and updates the
+/// internal package state.
+struct Subscriber<T: PackageDiscovery> {
+    exit_rx: oneshot::Receiver<()>,
+    filter: WorkspaceGlobs,
+    recv: broadcast::Receiver<Result<Event, NotifyError>>,
+    package_data: Arc<Mutex<HashMap<AbsoluteSystemPathBuf, RootWorkspaceData>>>,
+    manager: PackageManager,
+    repo_root: AbsoluteSystemPathBuf,
+    discovery: T,
+}
+
+impl<T: PackageDiscovery + Send + 'static> Subscriber<T> {
+    fn new(
+        exit_rx: oneshot::Receiver<()>,
+        repo_root: AbsoluteSystemPathBuf,
+        package_data: Arc<Mutex<HashMap<AbsoluteSystemPathBuf, RootWorkspaceData>>>,
+        recv: broadcast::Receiver<Result<Event, NotifyError>>,
+        discovery: T,
+    ) -> Result<Self, Error> {
+        let manager = PackageManager::get_package_manager(&repo_root, None)?;
+
+        Ok(Self {
+            exit_rx,
+            filter: manager.get_workspace_globs(&repo_root)?,
+            package_data,
+            recv,
+            manager,
+            repo_root,
+            discovery,
+        })
+    }
+
+    async fn watch(mut self) {
+        // initialize the contents
+        self.rediscover_packages().await;
+
+        // respond to changes
+        loop {
+            tokio::select! {
+                _ = &mut self.exit_rx => return,
+                file_event = self.recv.recv().into_future() => match file_event {
+                    Ok(Ok(event)) => self.handle_file_event(event).await,
+                    // if we get an error, we need to re-discover the packages
+                    Ok(Err(_)) => self.rediscover_packages().await,
+                    Err(RecvError::Closed) => return,
+                    // if we end up lagging, warn and rediscover packages
+                    Err(RecvError::Lagged(count)) => {
+                        tracing::warn!("lagged behind {count} processing file watching events");
+                        self.rediscover_packages().await;
+                    },
+                }
+            }
+        }
+    }
+
+    async fn handle_file_event(&mut self, file_event: Event) {
+        let glob_source_change = self.manager.workspace_glob_source(&self.repo_root);
```

Comment:
We should probably also watch `package.json` always as someone could alter `packageManager` where that causes us to invalidate `self.manager`. Maybe we also only use this strategy if that's set as it would be very confusing what to do if while we're watching a `pnpm-lock.yaml` and `yarn.lock` get created.